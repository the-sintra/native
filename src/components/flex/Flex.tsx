import React from 'react';
import { View, ViewStyle, ViewProps } from 'react-native';
import { Spacing } from '@the-sintra/core';
import { LayoutSizeProps } from '@/types/component';
import { FlexDirection, FlexJustify, FlexAlign, FlexWrap } from './Flex.type';
import { getGapValue } from './Flex.function';

export interface FlexProps extends LayoutSizeProps, ViewProps {
  direction?: FlexDirection;
  justify?: FlexJustify;
  align?: FlexAlign;
  wrap?: FlexWrap;
  gap?: keyof typeof Spacing | number;
}

export const Flex = React.memo(function Flex(props: FlexProps) {
  const {
    children,
    direction = FlexDirection.Row,
    justify = FlexJustify.Start,
    align = FlexAlign.Stretch,
    wrap = FlexWrap.Nowrap,
    gap,
    fullHeight,
    fullWidth,
    fitContent,
    style,
    ...rest
  } = props;

  const gapValue = getGapValue(gap);

  const flexStyles: ViewStyle = {
    flexDirection: direction,
    justifyContent: justify,
    alignItems: align,
    flexWrap: wrap,
    ...(fullHeight ? { height: '100%' } : {}),
    ...(fullWidth ? { width: '100%' } : {}),
    ...(fitContent ? { alignSelf: 'flex-start' } : {}),
    ...(gapValue !== undefined ? { gap: gapValue } : {}),
  };

  return (
    <View
      style={[flexStyles, style]}
      {...rest}
    >
      {children}
    </View>
  );
}); 