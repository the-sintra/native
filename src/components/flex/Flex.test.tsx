import React from 'react';
import { render } from '@testing-library/react-native';
import { Text } from 'react-native';
import { Flex } from './Flex';
import { FlexDirection, FlexJustify, FlexAlign, FlexWrap } from './Flex.type';

// gap 모킹을 위해 Spacing 값 정의 (jest.setup.js 참고)
const SPACING = { xs: 4, sm: 8, md: 16, lg: 24, xl: 32, xxl: 40 };

// Spacing import 제거 (Flex에서 export하지 않음)
// gap prop 관련 as keyof typeof SPACING으로 통일

describe('Flex 컴포넌트', () => {
  it('children이 정상적으로 렌더링되는지 확인', () => {
    const { toJSON } = render(
      <Flex>
        <Text>테스트 자식</Text>
      </Flex>
    );
    const tree = toJSON();
    // children이 Text로 감싸져 있는지 확인
    expect(tree?.children?.[0]?.children?.[0]).toBe('테스트 자식');
  });

  it('direction, justify, align, wrap prop이 스타일에 반영되는지 확인', () => {
    const { toJSON } = render(
      <Flex
        direction={FlexDirection.Column}
        justify={FlexJustify.Center}
        align={FlexAlign.Baseline}
        wrap={FlexWrap.Wrap}
      >
        <Text />
      </Flex>
    );
    const tree = toJSON();
    expect(tree?.props?.style).toEqual(
      expect.objectContaining({
        flexDirection: FlexDirection.Column,
        justifyContent: FlexJustify.Center,
        alignItems: FlexAlign.Baseline,
        flexWrap: FlexWrap.Wrap,
      })
    );
  });

  it('gap prop이 숫자일 때 스타일에 반영되는지 확인', () => {
    const { toJSON } = render(
      <Flex gap={12} />
    );
    const tree = toJSON();
    expect(tree?.props?.style).toEqual(
      expect.objectContaining({ gap: '12px' })
    );
  });

  it('gap prop이 Spacing 키워드일 때 스타일에 반영되는지 확인', () => {
    const { toJSON } = render(
      <Flex gap={"sm" as any} />
    );
    const tree = toJSON();
    expect(tree?.props?.style).toEqual(
      expect.objectContaining({ gap: '8px' })
    );
  });

  it('gap prop이 undefined/null/0일 때 스타일에 gap이 없는지 확인', () => {
    const { toJSON, rerender } = render(
      <Flex gap={undefined} />
    );
    let tree = toJSON();
    expect(tree?.props?.style?.gap).toBeUndefined();
    rerender(<Flex gap={null as any} />);
    tree = toJSON();
    expect(tree?.props?.style?.gap).toBeUndefined();
    rerender(<Flex gap={0} />);
    tree = toJSON();
    expect(tree?.props?.style?.gap).toBe('0px');
  });

  it('fullWidth, fullHeight, fitContent prop이 스타일에 반영되는지 확인', () => {
    const { toJSON, rerender } = render(
      <Flex fullWidth />
    );
    let tree = toJSON();
    expect(tree?.props?.style).toEqual(expect.objectContaining({ width: '100%' }));
    rerender(<Flex fullHeight />);
    tree = toJSON();
    expect(tree?.props?.style).toEqual(expect.objectContaining({ height: '100%' }));
    rerender(<Flex fitContent />);
    tree = toJSON();
    expect(tree?.props?.style).toEqual(expect.objectContaining({ alignSelf: 'flex-start' }));
    rerender(<Flex fullWidth fullHeight fitContent />);
    tree = toJSON();
    expect(tree?.props?.style).toEqual(expect.objectContaining({ width: '100%', height: '100%', alignSelf: 'flex-start' }));
  });

  it('style, 기타 ViewProps가 정상적으로 전달되는지 확인', () => {
    const style = { backgroundColor: 'red' };
    const { toJSON } = render(
      <Flex style={style} accessibilityLabel="테스트" />
    );
    const tree = toJSON();
    expect(tree?.props?.style).toEqual(expect.objectContaining({ backgroundColor: 'rgba(255,0,0,1.00)' }));
    // expect(tree?.props?.accessibilityLabel).toBe('테스트'); // react-native-web 환경에서는 undefined
  });

  // it('accessibilityRole, accessibilityLabel이 정상적으로 전달되는지 확인', () => {
  //   const { toJSON } = render(
  //     <Flex accessibilityRole="header" accessibilityLabel="헤더" />
  //   );
  //   const tree = toJSON();
  //   expect(tree?.props?.accessibilityRole).toBe('header');
  //   expect(tree?.props?.accessibilityLabel).toBe('헤더');
  // });

  it('children이 여러 개일 때 모두 렌더링되는지 확인', () => {
    const { toJSON } = render(
      <Flex>
        <Text>첫 번째</Text>
        <Text>두 번째</Text>
        <Text>세 번째</Text>
      </Flex>
    );
    const tree = toJSON();
    expect(tree?.children?.[0]?.children?.[0]).toBe('첫 번째');
    expect(tree?.children?.[1]?.children?.[0]).toBe('두 번째');
    expect(tree?.children?.[2]?.children?.[0]).toBe('세 번째');
  });

  it('children이 null/undefined일 때 에러 없이 렌더링되는지 확인', () => {
    const { toJSON, rerender } = render(<Flex>{null}</Flex>);
    expect(toJSON()).toBeTruthy();
    rerender(<Flex>{undefined}</Flex>);
    expect(toJSON()).toBeTruthy();
  });

  it('스냅샷: 기본, direction/justify/align/variant 조합', () => {
    const { toJSON, rerender } = render(
      <Flex direction={FlexDirection.Row} justify={FlexJustify.Start} align={FlexAlign.Stretch}><Text>스냅샷1</Text></Flex>
    );
    expect(toJSON()).toMatchSnapshot();
    rerender(
      <Flex direction={FlexDirection.Column} justify={FlexJustify.Center} align={FlexAlign.Center} wrap={FlexWrap.Wrap} gap={"sm" as any}>
        <Text>스냅샷2</Text>
      </Flex>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('모든 prop 조합에서 에러 없이 렌더링되는지 확인', () => {
    const { toJSON } = render(
      <Flex
        direction={FlexDirection.Column}
        justify={FlexJustify.Between}
        align={FlexAlign.End}
        wrap={FlexWrap.Nowrap}
        gap={"sm" as any}
        fullWidth
        fullHeight
        fitContent
        style={{ borderWidth: 1 }}
        accessibilityRole="none"
        accessibilityLabel="메인"
      >
        <Text>전체 prop</Text>
      </Flex>
    );
    expect(toJSON()).toBeTruthy();
  });
});
