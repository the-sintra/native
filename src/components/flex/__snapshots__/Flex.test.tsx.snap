// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flex 컴포넌트 스냅샷: 기본, direction/justify/align/variant 조합 1`] = `
<div
  className="css-view-175oi2r"
  dir={null}
  style={
    {
      "alignItems": "stretch",
      "flexDirection": "row",
      "flexWrap": "nowrap",
      "justifyContent": "flex-start",
    }
  }
>
  <div
    className="css-text-146c3p1"
    dir="auto"
  >
    스냅샷1
  </div>
</div>
`;

exports[`Flex 컴포넌트 스냅샷: 기본, direction/justify/align/variant 조합 2`] = `
<div
  className="css-view-175oi2r"
  dir={null}
  style={
    {
      "alignItems": "center",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "8px",
      "justifyContent": "center",
    }
  }
>
  <div
    className="css-text-146c3p1"
    dir="auto"
  >
    스냅샷2
  </div>
</div>
`;
